app/main.py

```diff
@@ -1,34 +1,133 @@
+        self.row = row
+        self.column = column
+        self.is_alive = is_alive
+
+    def hit(self) -> None:
+        self.is_alive = False
 
 
 class Ship:
-    def __init__(self, start, end, is_drowned=False):
-        # Create decks and save them to a list `self.decks`
-        pass
+    def __init__(
+            self,
+            start: Tuple[int, int], end: Tuple[int, int]) -> None:
+        self.start = start
+        self.end = end
+        self.decks: List[Deck] = []
+        self.is_drowned = False
+        self._create_decks()
 
-    def get_deck(self, row, column):
-        # Find the corresponding deck in the list
-        pass
+    def _create_decks(self) -> None:
+        if self.start[0] == self.end[0]:  # Horizontal ship
+            for col in range(self.start[1], self.end[1] + 1):
+                self.decks.append(Deck(self.start[0], col))
+        elif self.start[1] == self.end[1]:  # Vertical ship
+            for row in range(self.start[0], self.end[0] + 1):
+                self.decks.append(Deck(row, self.start[1]))
 
-    def fire(self, row, column):
-        # Change the `is_alive` status of the deck
-        # And update the `is_drowned` value if it's needed
-        pass
+    def fire(self, row: int, col: int) -> str:
+        # Check if this cell matches a deck
+        for deck in self.decks:
+            if deck.row == row and deck.column == col:
+                deck.hit()
+                if all(not d.is_alive for d in self.decks):
+                    self.is_drowned = True
+                return "Sunk!" if self.is_drowned else "Hit!"
+        return "Miss!"
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(
+            self,
+            ships: List[Tuple[Tuple[int, int], Tuple[int, int]]]) -> None:
+        self.field: dict[Tuple[int, int], Ship] = {}
+        self.ships: List[Ship] = []
+        self._validate_field(ships)
+        self._place_ships(ships)
+
+    def _place_ships(
+            self,
+            ships: List[Tuple[Tuple[int, int], Tuple[int, int]]]) -> None:
+        # Place the ships on the field and store them
+        for ship_coords in ships:
+            start, end = ship_coords
+            # First, check if the ship is adjacent to any other ship
+            if self._check_adjacent_cells(start, end):
+                raise ValueError("Ships cannot be in neighboring cells.")
+            ship = Ship(start, end)
```

Comment:
The adjacency check is performed after the ship is created. Consider performing this check before creating the ship to prevent incorrect placements.